example/notify_inval_entry.c

```diff
@@ -327,7 +324,7 @@ int main(int argc, char *argv[]) {
     update_fs();
 
     se = fuse_session_new(&args, &tfs_oper,
-                          sizeof(tfs_oper), NULL);
+                          sizeof(tfs_oper), &se);
```

Comment:
Dunno, yeah it works, but it makes the code complex? Use container_of() in tfs_init() to retrieve 'struct fuse_session' from 'struct fuse_conn_info'?

Reply:
There are copy-and-pasted `container_of` definitions in `fuse.c` and `fuse_lowlevel.c`:

```c
#define container_of(ptr, type, member) ({                              \
			const typeof( ((type *)0)->member ) *__mptr = (ptr); \
			(type *)( (char *)__mptr - offsetof(type,member) );})
```

I could copy-and-paste them into this file as well, and use them.  But this seems rather more complicated to me than using straight-forward pointers?  Especially since the `userdata` parameter seems to be designed for exactly the kind of 'passing-of-extra-info' that we are doing here?

In any case, I'm happy to use copy and use that macro, if you like it.  (Or we re-organise and stick it into a common internal header, even.)

If I understand that macro right, it would only work in `tfs_init` for as long as the `struct fuse_conn_info` we are getting is really a part of a `struct fuse_session`?  We'd need to document that this is now a requirement and something that the API would never be allowed to change?

Reply:
I guess the macro could go into fuse_common.h. I only yesterday realized libfuse already has it and had placed another copy into my uring patches. And yes, it only works as long as fuse_conn_info is part of fuse_session. From my point of view stopping correctly in the ->init() method is a common issue - I know of at least two other file system exiting there (one opensource is unionfs-fuse - I had placed an exit there  myself) and typically one stores local information into `userdata`, see for example passthrough_ll.c. 
So at least for me using `container_of` seems to be the better example. And then documenting that one can do that would be helpful to others as well. @Nikratio  what do you think?

Reply:
A `fuse session` instance is required not only to exit the session, but also for the various `fuse_notify_*` functions. So I think the problem of getting this instance is not limited to just the `init` callback.

I believe using `container_of` is common in kernel-code, but I'm not a friend of it at all since it leaks implementation details.

In this particular example, I think we can simply make `fuse_session` a global variable and be done with it.

Where global variables are not desirable, I think the `userdata` mechanism should be used for this. I would expect that in most cases this is a struct where a new session member can easily be added and carried around (this seems to apply to passthrough_ll.c as well).